test/Npgsql.Tests/DataSourceTests.cs

```diff
@@ -380,4 +381,86 @@ public async Task ConfigureJsonOptions_is_order_independent()
             Assert.That(reader.GetFieldValue<Test>(0).Id, Is.EqualTo(1));
         }
     }
+
+    [Test]
+    public async Task ReloadTypes()
+    {
+        using var adminConnection =  OpenConnection();
+        var type = await GetTempTypeName(adminConnection);
+
+        using var dataSource = CreateDataSourceWithType(type);
+        using var connection = dataSource.OpenConnection();
+
+        connection.ExecuteNonQuery($"CREATE TYPE {type} AS ENUM ('sad', 'ok', 'happy')");
+        dataSource.ReloadTypes();
+
+        Assert.Throws<InvalidCastException>(() => connection.ExecuteScalar($"SELECT 'happy'::{type}"));
+
+        // Close connection and reopen to make sure it picks up the new type and mapping from the data source
+        connection.Close();
+        connection.Open();
+
+        Assert.DoesNotThrow(() => connection.ExecuteScalar($"SELECT 'happy'::{type}"));
+    }
+
+    [Test]
+    public async Task ReloadTypesAsync()
```

Comment:
This test can be merged with the above, to exercise both sync and async in the same test (with a `[Values] bool async` parameter). Note that the only thing that needs to differ between the two tests is the invocation of ReloadTypes() vs. ReloadTypesAsync() - the rest of the code (e.g. the ExecuteScalarAsync()) can be async for both cases (that's not what we're testing here).

Reply:
Tests merged